packages/synapse-react-client/src/components/TableColumnSchemaEditor/TableColumnSchemaEditorUtils.ts

```diff
@@ -245,78 +258,95 @@ export function canHaveRestrictedValues(
-  let newSchema: SetOptional<ColumnModel, 'id'>[] = []
-  for (const m of proposedSchema) {
-    const copy: SetOptional<ColumnModel, 'id'> = {
-      ...m,
-    }
-    if (copy.id != null) {
-      // any changes to the existing column model?
-      const oldColumnModel: ColumnModel | undefined =
-        oldColumnModelId2Model.get(copy.id)
-      if (oldColumnModel != undefined && !isEqual(oldColumnModel, copy)) {
-        delete copy.id
+export function transformFormDataToColumnModels(
+  formData: ColumnModelFormData[],
+): SetOptional<ColumnModel, 'id'>[] {
+  return formData.map(
+    (formEntry: ColumnModelFormData): SetOptional<ColumnModel, 'id'> => {
+      // Remove the isSelected field because it was only used for the UI.
+      const { isSelected, ...rest } = formEntry
+      const columnModel = rest as SetOptional<ColumnModel, 'id'>
+      if (columnModel.jsonSubColumns) {
+        columnModel.jsonSubColumns = (
+          columnModel.jsonSubColumns as JsonSubColumnModelFormData[]
+        ).map(
+          (
+            jsonSubColumnFormData: JsonSubColumnModelFormData,
+          ): JsonSubColumnModel => {
+            // isSelected field from the subcolumn for the same reason
+            const { isSelected, ...rest } = jsonSubColumnFormData
+            return {
+              ...rest,
+            }
+          },
+        )
       }
-    }
-    newSchema.push(copy)
-  }
-  const createdColumnModels = (await createColumnModels(accessToken, newSchema))
-    .list
+      return columnModel
+    },
+  )
+}
 
-  // now that all columns have been created, figure out the column changes (create, update, and no-op)
-  // keep track of column ids to figure out what columns were deleted.
-  const changes: ColumnChange[] = []
-  const columnIds: Set<string | null> = new Set()
-  for (let i = 0; i < proposedSchema.length; i++) {
-    const oldColumnId: string | null = proposedSchema[i].id ?? null
-    const newColumnId: string | null = createdColumnModels[i].id ?? null
-    columnIds.add(oldColumnId)
-    columnIds.add(newColumnId)
-    if (oldColumnId !== newColumnId) {
-      changes.push({ oldColumnId, newColumnId })
+/**
+ * Transform ColumnModels returned by Synapse into the form data for the TableColumnSchemaForm.
+ * Default column models, if available, are used to determine which columns should not be editable in the form.
+ * @param columnModels
+ * @param defaultColumns
+ */
+export function transformColumnModelsToFormData(
+  columnModels: ColumnModel[],
+  defaultColumns: ColumnModel[] = [],
+): ColumnModelFormData[] {
+  return columnModels.map((cm): ColumnModelFormData => {
+    return {
+      ...cm,
+      isSelected: false,
+      // If the name matches a known default column model, then we consider it to be a default column model itself
+      isOriginallyDefaultColumn: defaultColumns.some(
+        dcm => dcm.name === cm.name,
```

Comment:
should we also check that the default column type also matches?

Reply:
Name alone is how SWC does it today: https://github.com/Sage-Bionetworks/SynapseWebClient/blob/6785747266c6e7177a94d03208eb95bcdc20e2e6/src/main/java/org/sagebionetworks/web/client/widget/table/v2/schema/ColumnModelsEditorWidget.java#L126-L128

Maybe a typical case to consider: if you deleted and re-added a default column `id` to be STRING instead of ENTITYID, should we still treat it as a default column?

Reply:
Probably a bug in SWC, but it's an edge case and I'm not going to push for a fix.
If your column name `id` with type STRING was added to an entity view, I suspect that the backend would not properly populate the column (since it's not really a default column), but I might be wrong!